### ISSUE TITLE
Convert custom training CSP-YOLOv4 weights to Tensorflow

### ISSUE THREAD

**Issue by RFedererno1** [AlexeyAB/darknet#7539]
Thank you for your contribution for community with this very good YOLO implementation.
I have converted yolov4-csp.weights file with yolov4-csp.cfg for COCO dataset you publiced to tensorflow and it worked fine.
In darknet:
![darknet_coco_result](https://user-images.githubusercontent.com/49802669/112435365-36814c00-8d77-11eb-8a40-9af92b5737e5.png)
And my weight convert implementation:
![tf_coco_result](https://user-images.githubusercontent.com/49802669/112435733-b7d8de80-8d77-11eb-9e36-350dba62a124.png)
As you can see, the results from darknet and tf with yolov4-csp.weights on COCO dataset are very close and they are accurate.
Then i trained yolov4-csp on my custom model on detecting face/facemask on darknet. Its result is very good on darknet detector. But after converting it to tf model, it cannot detect any object. The only change i made to config file is replace last filter from 255 = (5 + 80)*3 to 21 = (5 + 2)*3.
Is there any change to darknet source code ?
Thanks for your answer.

**Comment by AlexeyAB** [comment]
What TF implementation did you use?
Maybe in your implementation the number of classes are hardcoded in the Python code.

**Comment by RFedererno1** [comment]
TF version is 2.4.1. I use colab for converting.
In TF code, i changed number of filter in the last conv to 21. And when i get the weights from my custom model darknet weights, It got exactly the number of parameters in my TF model.
It found that the mish function in darknet is different from the real mish function. Can it cause the problem to model calculation ?

**Comment by AlexeyAB** [comment]
> It found that the mish function in darknet is different from the real mish function. Can it cause the problem to model calculation ?

This is the same Mish-function, just accelerated for CUDA.

Did you write your own code of Yolo for TF?
Can you share both versions of your code for 80 classes and 2 classes?

**Comment by RFedererno1** [comment]
Yes, i wrote my own model with the way of copying weight from darknet based on hunglc code.
About the code, the first part of model is identical for 80 class and 2 class.
The first change is the number of filter in the last conv:
80 classes:
![image](https://user-images.githubusercontent.com/49802669/112565701-1f426d00-8e10-11eb-971a-9b006db5f9b0.png)
2 classes:
![image](https://user-images.githubusercontent.com/49802669/112565724-2e291f80-8e10-11eb-8bf3-28e253c506f1.png)
Then i decoded the output of the model, this is for 80 classes and 2 classes:
![image](https://user-images.githubusercontent.com/49802669/112566012-c9ba9000-8e10-11eb-8685-a1c20db4b1f6.png)
Last, i nms the predictions:
80 classes:
![image](https://user-images.githubusercontent.com/49802669/112566203-1b631a80-8e11-11eb-8434-ca73c0e3fcdc.png)
2 classes:
![image](https://user-images.githubusercontent.com/49802669/112566319-4b122280-8e11-11eb-8ae6-dfa554ae9d6e.png)
Everything else is identical for both 80 classes and 2 classes.
For the link to my source code: https://colab.research.google.com/drive/1IaVXKKG12ghh7Iu3ZHIBig_fs3nR653c?usp=sharing

**Comment by RFedererno1** [comment]
This is 1 result i ran on darknet and TF with the same weight i trained on darknet:
darknet:
![image](https://user-images.githubusercontent.com/49802669/112568132-937f0f80-8e14-11eb-85a8-a62e8d672cd7.png)
It is pretty good.
TF:
![image](https://user-images.githubusercontent.com/49802669/112568168-a265c200-8e14-11eb-8345-35b471b9993e.png)
It can predict some of objects, but very bad accurate.
Maybe because the resize, RGB, of model calculation for floating foint ?
